src/lib/core/media-marshaller/print-hook.service.ts

```diff
@@ -0,0 +1,106 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import {Inject, Injectable} from '@angular/core';
+
+import {MediaChange} from '../media-change';
+import {BreakPoint} from '../breakpoints/break-point';
+import {LAYOUT_CONFIG, LayoutConfigOptions} from '../tokens/library-config';
+import {BreakPointRegistry, OptionalBreakPoint} from '../breakpoints/break-point-registry';
+
+/**
+ * Interface to apply PrintHook to call anonymous `target.updateStyles()`
+ */
+export interface HookTarget {
+  activatedBreakpoints: BreakPoint[];
+
+  updateStyles(): void;
+}
+
+@Injectable({providedIn: 'root'})
+export class PrintHookService {
+  protected offlineActivations: BreakPoint[] | null = null;
+
+  constructor(
+      @Inject(LAYOUT_CONFIG) protected layoutConfig: LayoutConfigOptions,
+      protected breakpoints: BreakPointRegistry,
+  ) {
+  }
+
+  /**
+   * Add 'print' mediaQuery: to listen for matchMedia activations
+   */
+  addPrintListener(queries: string[]) {
+    if (!!this.printAlias) {
+      queries.push('print');
+    }
+    return queries;
+  }
+
+  /**
+   * Is this service currently in Print-mode ?
+   */
+  get isPrinting(): boolean {
+    return !!this.offlineActivations;
+  }
+
+  /**
+   * What is the desired mqAlias to use while printing?
+   */
+  get printAlias(): string {
+    return this.layoutConfig.printWithBreakpoint || '';
+  }
+
+  /**
+   * Lookup breakpoint associated with print alias.
+   */
+  get printBreakPoint(): OptionalBreakPoint {
+    return this.breakpoints.findByAlias(this.printAlias!);
+  }
+
+  /**
+   * Prepare RxJs filter operator with partial application
+   * @return pipeable filter predicate
+   */
+  interceptEvents(target: HookTarget) {
+    return (change: MediaChange): boolean => {
+      if (change.mediaQuery == 'print') {
+
+        if (change.matches && !this.isPrinting) {
+          this.enablePrintMode(target, this.printBreakPoint);
+          target.updateStyles();
+        } else if (!change.matches && this.isPrinting) {
+          this.disablePrintMode(target);
+          target.updateStyles();
```

Comment:
I very much dislike this as it seems to violate a core tenet of functional programming, namely that it produces a side effect. Can we rework this into a different workflow that isolates this better?